File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        lengthNodeA = 0
        lengthNodeB = 0
        tempA = headA
        tempB = headB
        while tempA:
            lengthNodeA += 1
            tempA = tempA.next
        while tempB:
            lengthNodeB += 1
            tempB = tempB.next
        while lengthNodeA > lengthNodeB:
            lengthNodeA -= 1
            headA = headA.next
        while lengthNodeB > lengthNodeA:
            lengthNodeB -= 1
            headB = headB.next
        while headA and headB:
            if headA == headB:
                return headA
            headA = headA.next
            headB = headB.next
        return None
        
        
            
        

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int lengthA = 0;
        int lengthB = 0;
        ListNode* tempA = headA;
        ListNode* tempB = headB;
        while(tempA){
            lengthA += 1;
            tempA = tempA->next;
        }
        while(tempB){
            lengthB += 1;
            tempB = tempB->next;
        }
        if(lengthA > lengthB){
            while(lengthA > lengthB){
                headA = headA->next;
                lengthA -= 1;
            }
        }
        if(lengthB > lengthA){
            while(lengthB > lengthA){
                headB = headB->next;
                lengthB -= 1;
            }
        }
        while(headA && headB){
            if(headA == headB){
                return headA;
            }
            headA = headA->next;
            headB = headB->next;
        }
        return 0;
        
    }
};